core/test/utils/matrix_generator.hpp

```diff
@@ -229,19 +233,34 @@ std::unique_ptr<MatrixType> generate_random_lower_triangular_matrix(
         // select a subset of `nnz_in_row` column indexes, and fill these
         // locations with random values
         std::shuffle(begin(col_idx), end(col_idx), engine);
-        std::for_each(begin(col_idx), begin(col_idx) + nnz_in_row,
-                      [&](size_type col) {
-                          if (col <= row) {
-                              if (ones_on_diagonal && col == row) {
-                                  data.nonzeros.emplace_back(row, col, one);
-                              } else {
-                                  data.nonzeros.emplace_back(
-                                      row, col,
-                                      detail::get_rand_value<value_type>(
-                                          value_dist, engine));
-                              }
-                          }
-                      });
+        // add non-zeros
+        bool has_diagonal{};
+        for (size_type nz = 0; nz < nnz_in_row; ++nz) {
+            auto col = col_idx[nz];
+            // skip non-zeros outside triangle
+            if (col > row && lower_triangular ||
+                col < row && !lower_triangular) {
+                continue;
+            }
+
+            // generate and store non-zero
+            auto val = detail::get_rand_value<value_type>(value_dist, engine);
+            if (col == row) {
+                has_diagonal = true;
+                if (ones_on_diagonal) {
+                    val = one;
+                }
+            }
+            data.nonzeros.emplace_back(row, col, val);
+        }
+
+        // add diagonal if it hasn't been added yet
+        if (!has_diagonal) {
+            auto val = ones_on_diagonal ? one
+                                        : detail::get_rand_value<value_type>(
+                                              value_dist, engine);
+            data.nonzeros.emplace_back(row, row, one);
```

Comment:
Pretty sure you mean:
```suggestion
            data.nonzeros.emplace_back(row, row, val);
```

Reply:
Good catch, thanks! Since we don't exactly test for non-one diagonal entries, this would have been hard to find.